doc/decisions/private_api.md

```diff
@@ -0,0 +1,52 @@
+# Library Split
+
+## Problem
+
+Only `libelektra-core` is supposed to access private data but this contradicts the goal to keep the library minimal.
+`kdbprivate.h` was too generic, it contained many other parts next to the struct definitions of Key/KeySet.
+
+## Constraints
+
+## Assumptions
+
+- The [C99 standard, section 5.2.4.1](http://www.open-std.org/jtc1/sc22/wg14/) gives following limit:
+  4095 external identifiers in one translation unit
+
+## Considered Alternatives
+
+- keep current situation
+
+## Decision
+
+- Also allow other libraries (e.g. a new `libelektra-operations`) to access to non-public API.
+- Put struct definitions of Key/KeySet in a separate header file, which gets included by parts that need it (see also [Header File Structure](header_file_structure.md)).
+- Define the current API version of the core API as
+  ```c
+  extern const int ELEKTRA_CORE_VERSION_MAJOR;
+  extern const int ELEKTRA_CORE_VERSION_MINOR;
+  extern const int ELEKTRA_CORE_VERSION_PATCH;
+  ```
+  in the public API.
+- Plugins can define compatible ranges of core API version via the contract. This version is checked during `elektraPluginOpen` by `libelektra-kdb`.
+
+## Rationale
+
+- allows various users (plugins, applications) to link to (more or less) exactly what they need
+- allows separate versioning for every library
+- allows alternative implementation of parts of Elektra, e.g. a libcore written in Rust
+- facilitates code reuse between plugins
+- The `extern const int` API version can be used for compatibility checks.
```

Comment:
But `system:/elektra/version/constants/KDB_VERSION` does not and cannot come from `libelektra-core`. The core library has no knowledge of the KDB and therefore cannot add a Key to it.

Also the name `KDB_VERSION` clearly suggests that this is the version of `libelektra-kdb` and not `libelektra-core`.

Reply:
I think we should have clear restrictions which libs can be used with which versions of our other libs. `kdb`, `mmap` and `core` always must have the same version, so no separate public identifiers are needed. (Maybe some other identifiers so that we can check ourselves if libs are consistent)

Reply:
1. I'd would use the name `ELEKTRA_VERSION`, if it is not just for `libelektra-kdb`.
2. The fact that the `system:/elektra` key can only be used in `libelektra-kdb`, global plugins and applications (i.e. the parts that have access to the full KDB), still stands. Only a `extern const int` in `libelektra-core` can be accessed everywhere and is guaranteed to come from the library code (`.so`) and from some header that may be edited/replaced.

If your concern is just about the use of 3 separate symbols, we could combine the whole thing into a single symbol:

```c
const uint32_t ELEKTRA_VERSION = 1002003; // -> 1.2.3
```

This allows more than 1000 major versions each with 1000 minor versions, which in turn could have 1000 patch versions each. That should be enough and would only be one single symbol.

For convenience we could define the macros:

```c
#define ELEKTRA_MAJOR_VERSION = (uint32_t)((ELEKTRA_VERSION) / 1000000);
#define ELEKTRA_MINOR_VERSION = (uint32_t)(((ELEKTRA_VERSION) / 1000) % 1000);
#define ELEKTRA_PATCH_VERSION = (uint32_t)((ELEKTRA_VERSION) % 1000);
```

Reply:
No, my concern is that we are getting lost in irrelevant stuff. Nobody asked for something like this, so why bother? In particular I do not see how it supports this decision or solves the problem written in the beginning of the decision.

Reply:
If your proposal is actually to rename KDB_VERSION (in kdb.h) to ELEKTRA_VERSION and use const uint32_t instead of macros, I am all in favor but this is a very different decision. The same for KDB_PATH_SEPARATOR, KDB_PATH_ESCAPE, KEY_NULL etc.

About the number of the symbols: please do a test how much these symbols cost in terms of library size. Guessing that it could be expensive and then do optimizations based on guesses doesn't sound very scientific ;)

Reply:
> No, my concern is that we are getting lost in irrelevant stuff. Nobody asked for something like this, so why bother? In particular I do not see how it supports this decision or solves the problem written in the beginning of the decision.

The problem this constant solves is that currently there is no way of knowing what version of the API is actually implemented by a `libelektra-core.so` file. Like I also explained [above](https://github.com/ElektraInitiative/libelektra/pull/4223#discussion_r783356201), the `#define`s we have are not good enough.

> If your proposal is actually to rename KDB_VERSION (in kdb.h) to ELEKTRA_VERSION and use const uint32_t instead of macros, I am all in favor but this is a very different decision.

Purely renaming `KDB_*` to `ELEKTRA_*` is a different decision, yes. The type change might be as well. However, the deciding that some kind of version constant should exist within the `libelektra-core.so` definitely belongs here. This decision is concerned with how and why (some) internal structures/APIs may be accessed from outside of `libelektra-core.so`. This can only be done safely, if we know what version of `libelektra-core.so` is used. This in turn requires some sort of version information within `libelektra-core.so`.

> About the number of the symbols: please do a test how much these symbols cost in terms of library size. Guessing that it could be expensive and then do optimizations based on guesses doesn't sound very scientific ;)

I was never really concerned about the number of symbols. You were the one that [brought it up](https://github.com/ElektraInitiative/libelektra/pull/4223#discussion_r783101589).


---

To sum this up: Can we agree on the following?

1. We replace 

    ```c
    #define KDB_VERSION "@KDB_VERSION@"
    #define KDB_VERSION_MAJOR @KDB_VERSION_MAJOR@
    #define KDB_VERSION_MINOR @KDB_VERSION_MINOR@
    #define KDB_VERSION_PATCH @KDB_VERSION_PATCH@
    ```

    with

    ```c
    extern const uint32_t ELEKTRA_VERSION; // defined in libelektra-core.so
    #define ELEKTRA_MAJOR_VERSION = (uint32_t)((ELEKTRA_VERSION) / 1000000);
    #define ELEKTRA_MINOR_VERSION = (uint32_t)(((ELEKTRA_VERSION) / 1000) % 1000);
    #define ELEKTRA_PATCH_VERSION = (uint32_t)((ELEKTRA_VERSION) % 1000);
    ```
2. I won't make this a separate decision file, because writing a whole document for this very small change seems very silly.

Reply:
> I was never really concerned about the number of symbols. You were the one that https://github.com/ElektraInitiative/libelektra/pull/4223#discussion_r783101589.

Yes, because once I greatly reduced the size of libelektra-core by reducing symbols. Please **first** benchmark what exactly we need to do to not bloat it again, and then let us design a proper API. If the `extern const uint32_t` are small, several one of them is a better API than macros for calculation.

Reply:
I didn't even know, if such a symbol would be of interest at all. Therefore, starting benchmarks for something that maybe isn't wanted at all, would be a waste of time.

I will do some benchmarks, but in this case, I don't think a benchmark is gonna tell us a lot. A `uint32_t` symbol should always be exactly 4 bytes. With function symbols this is of course different. Whether or not a function is exported affects what optimizations the compiler can do, so the code-size of the function may change drastically. But AFAIK (other than possible alignment reasons) whether or not a `uint32_t` is exported can only ever make a difference of 4 bytes.